Add new package: sgpp

G-071 opened the pull request.

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    maintainers = ['G-071', 'leiterrl', 'pfluegdk']
+
+    # Versions with Python 3 bindings:
+    version('master', git='https://github.com/SGpp/SGpp.git', branch='master')
```
You can actually move this `git` up above below `url`

G-071 replied:
Fixed in dcfd800

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    # Mandatory dependencies
+    depends_on('scons@2.5.1', when='@:3.1.0', type=('build'))
+    depends_on('scons@3:', when='@3.2.0:', type=('build'))
+    depends_on('zlib', type=('build', 'run'))
```
This deptype is definitely wrong, as zlib doesn't install any binaries. I think you want type=link here.

G-071 replied:
Fixed in dcfd800

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    depends_on('py-numpy@:1.16', when='@:3.1.0+python', type=('build', 'run'))
+    depends_on('py-numpy@1.17:', when='@3.2.0:+python', type=('build', 'run'))
+    depends_on('py-scipy@:1.2.3', when='@:3.1.0+python', type=('build', 'run'))
+    depends_on('py-scipy@1.3.0:', when='@3.2.0:+python', type=('build', 'run'))
```
Previous versions of numpy and scipy also support Python 3. I would just list the minimum version and let the concretizer handle the rest.

G-071 replied:
I thought this would work
```python
    depends_on('py-numpy@1.16.0:', when='+python', type=('build', 'run'))
    depends_on('py-scipy@1.2.3:', when='+python', type=('build', 'run'))
```
as the minimum listed versions work for both Python 2 and 3. However, if I try this with the Python 2 version of sgpp ```spack spec sgpp@3.1.0```, it still tries to use the newest numpy (1.18.2) instead of one that still supports Python 2 yielding the error ```py-numpy requires python version 3.5:, but spec asked for 2.7.16```

Fixing the version to an older one like this:
```python
    depends_on('py-numpy@1.16.0', when='+python', type=('build', 'run'))
    depends_on('py-scipy@1.2.3', when='+python', type=('build', 'run'))
```
would work for both Python 3 and Python 2. 

That being said: Unless I am missing any downsides to it, I would prefer to simply leave it as it is now! This way we will always get the newest numpy version for the more current SGpp versions and only use the older numpy versions for the old, python2-versions of SGpp.

adamjstewart replied:
Yeah, that's a bug in Spack's concretizer that will be fixed when the new concretizer is released. I'm fine with leaving it as is until that bug has been fixed.

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    variant('simd',
+            default='sse3',
+            values=('sse3', 'sse42', 'avx', 'fma4', 'avx2', 'avx512'),
+            description='Specifies the SIMD extension to be used')
```
@alalazo @becker33 can one of you explain how to use Spack's microarch support to detect this automatically?

alalazo replied:
Sure, this is documented [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#architecture-specifiers). In short, you can code conditionals like:
```python
if 'sse3' in self.spec.target:
    # do something
```
within methods called during install and take appropriate actions in each branch.

G-071 replied:
I have removed the simd variant in c59f5a6 and now use the information in the spec to set the scons flag appropriately.

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    conflicts('+combigrid', when='@:3.2.0~quadrature')
+
+    def build_args(self, spec, prefix):
+        # No need for unit tests anymore -> saves installation time
```
If you want, you can actually make things conditional based on whether or not tests are requested:
```python
if self.run_tests:
    self.args = [...]
else:
    self.args = [...]
```
Then tests can be built when a user runs `spack install --test=root sgpp`

G-071 replied:
That's interesting - I will definitely add that! Thanks!

G-071 replied:
Added in 2225495

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+                     'RUN_BOOST_TESTS=0',
+                     'RUN_PYTHON_TESTS=0']
+        # No need to check the code style anymore -> saves installation time
+        if any(x in spec for x in ['@3.2.0', '@3.1.0', '@3.0.0']):
```
This can be simplified like so:
```python
if spec.satisfies('@:3.2.0'):
```

G-071 replied:
Added in dcfd800

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    # Fix faulty setup.py in 3.2.0
+    patch('fix-setup-py.patch', when='@3.2.0')
+    # Backport opencl fix
+    patch('ocl.patch', when='@:3.2.0+opencl')
```
Have these patches been submitted upstream?

G-071 replied:
The patches have been submitted upstream and are part of sgpp in the versions 3.3.0 and newer (master).

However, I have talked with the current maintainer and there are currently no plans to create updated versions of the old releases. Hence, to build the sgpp spack package with the old versions, these patches will remain relevant for the spack package!

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    # Java dependencies
+    depends_on('swig@3:', when='+java', type=('build'))
+    depends_on('openjdk', when='+java', type=('build', 'run'))
+    # Python libraries (version depends on whether we use Python 2 or 3)
```
You're missing a build dependency on `py-setuptools`

G-071 replied:
Fixed in dcfd800

adamjstewart commented on var/spack/repos/builtin/packages/sgpp/package.py:
```diff
@@ -0,0 +1,167 @@
+    variant('java', default=False,
+            description='Provide Java bindings for SGpp')
+    # export CLASSPATH=$(spack location --install-dir sgpp)/lib/jsgpp.jar
+    # After that one can compile and run simple java examples with sgpp
```
Spack should automatically set `CLASSPATH` if you load the generated module or use `spack load` to load the package.

G-071 replied:
This doesn't seem to work with the current version of this sgpp package. 

I played around with it a bit today and found out that I can get it to work if I use ```extends('openjdk', when='+java')``` instead of ```depends_on```. 
However, that alone is not sufficient. I also had to remove the line ```extends('python', when='+python')```, and only then the ```CLASSPATH``` was set automatically by ```spack load sgpp``` after installing it with ```spack install sgpp -python +java```.

Is this related to #987, or did I mess up somewhere?

adamjstewart replied:
Yes, that's the same bug. You need to `extends('java', when='+java')`, but it breaks when a package extends multiple other packages.

G-071 replied:
Is there any way around this bug?

Otherwise I would just remove the Java variant from the package until the bug is fixed. We have more people using the Python bindings, so I would favor those over the Java ones.

G-071 replied:
I removed the java extension from the package in a6c9c67

adamjstewart replied:
You have a few options:

1. Don't support Java bindings, no variant and no dep
2. Support it but without env vars set, just a dep but not an extends
3. Support it correctly, use dep but no extends, manually set env vars

G-071 commented:
I think I have addressed the requested changes from above!
@adamjstewart Could you have another look at it please? I'd be happy make more changes if necessary! 

Edit: Note that I added one more patch for a bug in SGpp I found whilst editing the package. As with the other patches it affects whether SGpp builds at all in a certain configuration, so I have included it. This patch will be in upstream as soon as SGpp/SGpp#229 is merged. This currently blocked because the cluster that runs the meta-testing check is down. The patch should be merged in its current form tomorrow, as soon as that runs again.
